### PR TITLE
Added possibility to use multiple attributes for static widgets

### LINKED ISSUE #757: Add option to show multiple attributes in entity state widget
**Is your feature request related to a problem? Please describe.**
Currently only one attribute can be shown in entity state widget.

**Describe the solution you'd like**
Option to use multiple attributes in entity state widget.

### PR DESCRIPTION
**Pull request by klejejs**
Fixes #757

### CODE REVIEW

**Comment by JBassett** [comment]
So before I review what are our thoughts about https://github.com/home-assistant/android/pull/790?  I know it isn't exactly what was asked for but I think it covers the use case and more.

**Comment by klejejs** [comment]
> So before I review what are our thoughts about #790? I know it isn't exactly what was asked for but I think it covers the use case and more.

I think your pull request gives a really nice addition to the app. At first I thought I should close mine, because, as you said, your pull request covers the use case and has many more possibilities, however, when I thought more about it, I decided that in my opinion we need it. I would be very happy to use template widget, as I could setup much more information in one widget than before, however, I know for a fact, I have family members who would never understand the templating part and I would never want them to. It would be really hard for them to set up a widget with some simple information.
TLDR: I love templating widget, but I also feel we should provide a more user friendly way to setup a state widget for less advanced users.

**Comment by klejejs** [comment]
> Ended up having a minute to review. Couple of things I noticed. Still looking to see if we want to expand this Widget any further or maybe just make it more focused on just state rather than attributes too. If people want more complex they can use the template widget. Thoughts?

Are you worried about code and functionality repetition. If yes, we can close this pull request and think how to improve templating widget. The problem I see is that templating widget simply gives user an empty box, this allows users to go through everything step by step with search functionality. Much better user experience. But that is just my opinion. :) Maybe someone has any other opinions?

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt [hunk 1]
```diff
@@ -172,10 +176,11 @@ class StaticWidget : AppWidgetProvider() {
             staticWidgetDao.add(StaticWidgetEntity(
                 appWidgetId,
                 entitySelection,
-                attributeSelection,
+                attributeSelection?.joinToString(","),
                 labelSelection,
                 textSizeSelection?.toFloatOrNull() ?: 30F,
-                if (separatorSelection == "") " " else separatorSelection ?: " "
+                stateSeparatorSelection ?: "",
+                attributeSeparatorSelection ?: ""
```
[the remark this answers is not in the document]
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt:183]: Correct. All of the previous widgets will migrate to have a space as a separator, however, new ones will have nothing, if user chooses "No separator" and doesn't type anything in the field. I felt this was a better approach as it allows to, for example, add `unit_of_measurement` right after the value like this "1.24‎€", previously it would have been like this "1.24 ‎€". Besides users still have an option to add space in the separator field.

File: app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt [hunk 2]
```diff
@@ -74,5 +78,34 @@ abstract class AppDatabase : RoomDatabase() {
                 database.execSQL("ALTER TABLE `static_widget` ADD `separator` TEXT NOT NULL DEFAULT ' '")
             }
         }
+        private val MIGRATION_4_5 = object : Migration(4, 5) {
+            override fun migrate(database: SupportSQLiteDatabase) {
+                try {
+                    val widgets = database.query("SELECT * FROM `static_widget`")
+                    widgets.use {
+                        if (widgets.moveToFirst()) {
```
[the remark this answers is not in the document]
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt]: You are right. Fixed that.

File: app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt [hunk 3]
```diff
@@ -85,5 +89,42 @@ abstract class AppDatabase : RoomDatabase() {
...
+                            database.execSQL("DROP TABLE IF EXISTS `static_widget`")
+                            database.execSQL("CREATE TABLE IF NOT EXISTS `static_widget` (`id` INTEGER NOT NULL, `entity_id` TEXT NOT NULL, `attribute_ids` TEXT, `label` TEXT, `text_size` FLOAT NOT NULL DEFAULT '30', `state_separator` TEXT NOT NULL DEFAULT '', `attribute_separator` TEXT NOT NULL DEFAULT '', PRIMARY KEY(`id`))")
+                        }
+                    }
+                    widgets.close()
+                } catch (exception: SQLiteException) {
+                    Log.e(exception.toString(), "SQLiteException in migrate from database version 4 to version 5")
+                } catch (exception: Exception) {
+                    Log.e(exception.toString(), "Failed to migrate database version 4 to version 5")
+                }
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt:126]: If we are doing the same thing no need for multiple catches.  Don't forget to create a TAG and pass the exception to the logger.
```
} catch (exception: Exception) {
    Log.e(TAG, "Failed to migrate database version 5 to version 6", exception)
}
```

Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt]: My bad. Didn't know this was the way to properly log exceptions. Fixed it.
